### PR TITLE
[TASK-9] As a user, I can sign in using email and password

### PR DESCRIPTION
**Pull request by rafayet-monon**
Task Card - https://github.com/rafayet-monon/surveyor/projects/1#card-43745565

## What happened
Users can sign in with email and password.
 
## Insight
- [x] Create login forms
- [x] Connect backend API
- [x] Write unit tests
- [x] Write UI tests
 
Users can sign in with valid email and password. Invalid email/password will show error along with validation errors.

## Proof Of Work

- Invalid Email or password

image

- Validation Error

image

- User Login

signin

### CODE REVIEW

File: cypress/integration/login.spec.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+const autoRecord = require('cypress-autorecord');
+
+describe('The Login Page', () => {
```
Comment (Lahphim) [inline cypress/integration/login.spec.js:3]: @rafayet-monon Should we describe all these like this below?
```
- The login page.
  - Logins with a valid credential.
    - Redirects to the dashboard page.
    - Shows a successful login message.
  - Logins with an invalid credential.
    - does NOT redirect to another page.
    - Shows an error message.
```

File: package.json [hunk 2]
```diff
@@ -14,7 +14,8 @@
     "stylelint": "stylelint '**/*.scss'",
     "lint-in-file": "eslint \"src/**/*.{js,jsx}\" --quiet --output-file ./eslint-output.json -f json",
     "danger": "npx danger ci",
-    "cypress": "cypress open"
+    "cypress-cli": "cypress open",
```
Comment (Lahphim) [inline package.json]: @rafayet-monon Should we use `cypress-open` and `cypress-run` or some other keys? 🤔

File: src/screens/Login/handler.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          }
+        })
+        .then((resData) => {
+          dispatch({
+            type: 'AUTH',
+            payload: resData.data,
+          });
+        });
+    } catch (error) {
+      if (error.response.status === 400) {
```
Comment (Lahphim) [inline src/screens/Login/handler.js:34]: @rafayet-monon Should it return with `401`? 🤔

File: cypress/integration/login.spec.js (hunk 1 above)
Comment (rafayet-monon) [reply cypress/integration/login.spec.js:3]: Changed in [7aeb592](https://github.com/rafayet-monon/surveyor/pull/14/commits/7aeb592db449371eace4b5a879592937ad8efbc8). But as there's no `successful login message` in design, that particular test is not available.

File: package.json (hunk 2 above)
Comment (rafayet-monon) [reply package.json]: Changed in [7aeb592](https://github.com/rafayet-monon/surveyor/pull/14/commits/7aeb592db449371eace4b5a879592937ad8efbc8).

File: src/screens/Login/handler.js (hunk 3 above)
Comment (rafayet-monon) [reply src/screens/Login/handler.js:34]: The api returns 400 in this case -

image

File: src/utils/Api.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import axios from 'axios';
+
+const authorization_token = localStorage.getItem('authorization_token');
+
+const instance = axios.create({
+  baseURL: process.env.REACT_APP_BACKEND_URL,
+  headers: {
+    'Access-Control-Allow-Origin': '*',
```
Comment (namtx) [inline src/utils/Api.js]: This header is repsonse header, not a request header according to https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS, no?

File: src/utils/Api.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    'Content-Type': 'application/json',
+    'X-Requested-With': 'XMLHttpRequest',
```
Comment (namtx) [inline src/utils/Api.js]: Do we need this header?

**Comment by namtx** [comment]
React_App
I get this error when trying to login

**Comment by rafayet-monon** [comment]
@namtx Fixed in [c5096c1](https://github.com/rafayet-monon/surveyor/pull/14/commits/c5096c1de2d1471f49938a826d37a5213cb2134d)

**Review by namtx — APPROVED** [review]

**Review by Lahphim — APPROVED** [review]

**Pull request merged**